Libplanet/Blockchain/BlockChain.cs

```diff
@@ -84,57 +89,59 @@ Block<T> genesisBlock
         {
         }
 
+        public BlockChain(
+            IBlockPolicy<T> policy,
+            IStore store,
+            HashDigest<SHA256> genesisHash
+        )
+            : this(policy, store, store.GetCanonicalChainId() ?? Guid.NewGuid(), genesisHash)
+        {
+        }
+
         internal BlockChain(
             IBlockPolicy<T> policy,
             IStore store,
             Guid id,
             Block<T> genesisBlock
         )
-            : this(
-                policy,
-                store,
-                id,
-                genesisBlock,
-                false
-            )
+            : this(policy, store, id, genesisBlock.Hash)
         {
+            if (Count == 0)
+            {
+                bool inFork = Store.ContainsBlock(genesisBlock.Hash);
+                Append(
+                    genesisBlock,
+                    currentTime: genesisBlock.Timestamp,
+                    renderActions: !inFork,
+                    evaluateActions: !inFork
+                );
+            }
         }
 
-        private BlockChain(
+        internal BlockChain(
             IBlockPolicy<T> policy,
             IStore store,
             Guid id,
-            Block<T> genesisBlock,
-            bool inFork
+            HashDigest<SHA256> genesisHash
         )
         {
             Id = id;
             Policy = policy;
             Store = store;
 
+            _genesisHash = genesisHash;
             _blocks = new BlockSet<T>(store);
             _transactions = new TransactionSet<T>(store);
             _rwlock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
             _txLock = new object();
 
-            if (Store.GetCanonicalChainId() is null)
+            var canonicalId = Store.GetCanonicalChainId();
+            if (canonicalId is null)
             {
                 Store.SetCanonicalChainId(Id);
             }
-
-            _logger = Log.ForContext<BlockChain<T>>()
-                .ForContext("CanonicalChainId", Id);
-
-            if (Count == 0)
-            {
-                Append(
-                    genesisBlock,
-                    currentTime: genesisBlock.Timestamp,
-                    renderActions: !inFork,
-                    evaluateActions: !inFork
-                );
-            }
-            else if (!Genesis.Equals(genesisBlock))
+            else if (Store.CountBlocks() > 0
+                     && !Store.IterateIndexes((Guid)canonicalId).First().Equals(_genesisHash))
```

Comment:
Can we use `IterateBlockHashes(limit=1)` instead of `IterateIndexes`?

Reply:
`IterateBlockHashes()` seems having no parameters. Also I'm not sure `IterateBlockHashes` to return in ascending order.

Reply:
I meant using `BlockChain<T>.IterateBlockHashes()` method, not `Store.IterateBlockHashes` method. `BlockChain<T>.IterateBlockHashes()` uses `Store.IterateIndexes` internally, so I think the order will be the same.